File: src/com/anathema_roguelike/characters/stats/resources/ResourceModification.java
/*******************************************************************************
 * This file is part of AnathemaRL.
 *
 *     AnathemaRL is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     AnathemaRL is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with AnathemaRL.  If not, see <http://www.gnu.org/licenses/>.
 *******************************************************************************/
package com.anathema_roguelike.characters.stats.resources;

import com.anathema_roguelike.characters.effects.Calculation;
import com.anathema_roguelike.characters.effects.Duration;
import com.anathema_roguelike.characters.effects.FixedCalculation;
import com.anathema_roguelike.characters.effects.FixedDuration;
import com.anathema_roguelike.characters.effects.buffs.Buff;
import com.anathema_roguelike.characters.effects.modifiers.Modifier;

public abstract class ResourceModification extends Buff {
	
	public ResourceModification(Object source, Class<? extends Resource> resource, int amount) {
		super(source, new FixedDuration(Duration.ROUND, 0));
		
		addModifier(new Modifier(source, resource, new FixedCalculation<Integer>(amount), null));
	}
	
	public ResourceModification(Object source, Class<? extends Resource> resource, Calculation<Integer> calculation) {
		super(source);
		
		addModifier(new Modifier(source, resource, calculation, null));
	}
}

File: src/com/anathema_roguelike/stats/characterstats/resources/ResourceModification.scala
/*******************************************************************************
 * Copyright (c) 2019. Carl Minden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

package com.anathema_roguelike
package stats.characterstats.resources

import com.anathema_roguelike.actors.Duration
import com.anathema_roguelike.entities.characters.Character
import com.anathema_roguelike.stats.effects.{Calculation, Effect, FixedCalculation, HasEffect}
import scala.reflect.runtime.universe._

class ResourceModification[T <: Resource : TypeTag](
      initiator: Option[Character],
      source: Option[HasEffect[Effect[Character, T]]],
      calculation: Calculation
  ) extends Effect[Character, T](source, List(), Duration.INSTANT) {

  def this(initiator: Option[Character], source: Option[HasEffect[Effect[Character, T]]], amount: Int) {

    this(initiator, source, new FixedCalculation(amount))
  }

  def getResource: Class[T] = typeTagToClass
  def getInitiator: Option[Character] = initiator

  override def onApplicationCallback(character: Character): Unit = {
    super.onApplicationCallback(character)
    character.modifyResource[T](getInitiator, getSource, calculation().intValue)
  }
}
